### PR TITLE
Refactor CLI, fix tests

### PR DESCRIPTION
**Pull request by schneems**
- The CLI class has been refactored so it can be tested.
  - The `Rundoc.project_root=` config option is being removed in favor of directly allowing file manipulation via ruby code in the `after_build` blocks.
  - Integration tests were not being executed, that's fixed.
  - Added an integration test for screenshots
  - Remove `rundoc.depend_on` in favor of `::-- rundoc.require`

### CODE REVIEW

**Comment by lilacstella** [comment]
- Should we mention that there is no more default source document in the README or the CHANGELOG? (this feels like losing backwards compatibility and thus a major version bump)
- when I run just `rundoc build`, there is a ruby error instead of the normal shell help command. 
- I am running into a bug with the success and fail directory name command argument where it's complaining about some string not having the exist method.

**Comment by lilacstella** [comment]
When I run 
`rundoc build`
I get a ruby error instead of a nice shell help message.

```
$ bin/rundoc build
bin/rundoc:59:in `exist?': no implicit conversion of nil into String (TypeError)

    if !File.exist?(@path)
                    ^^^^^
	from bin/rundoc:59:in `build'
	from .../ruby/3.3.2/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from .../ruby/3.3.2/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from .../ruby/3.3.2/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from .../ruby/3.3.2/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from bin/rundoc:76:in `<main>'
```

**Comment by lilacstella** [comment]
Whenever I use the `--on-successful-dir` and `--on-failure-dir` arguments, I get ruby errors. 

`$ bin/rundoc build --path=rundoc.md --on-success-dir=yay --on-failure-dir=sad
## Running your docs
## No .env file found /Users/stella.yang/Documents/rundoc/.env, skipping dotenv loading
## Clearing on failure directory sad
## Cleaning working directory /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240711-18299-vms2hf
/Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:121:in `ensure in call': undefined method `exist?' for an instance of String (NoMethodError)

        FileUtils.remove_entry_secure(execution_context.output_dir) if on_failure_dir.exist?
                                                                                     ^^^^^^^
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:121:in `call'
	from bin/rundoc:72:in `build'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from bin/rundoc:76:in `<main>'
/Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:93:in `call': undefined method `exist?' for an instance of String (NoMethodError)

      FileUtils.remove_entry_secure(on_failure_dir) if on_failure_dir.exist?
                                                                     ^^^^^^^
	from bin/rundoc:72:in `build'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
```

The rundoc.md is a hello world document
````
hi

```
:::>> $ echo 'hi'
```
````

**Comment by lilacstella** [comment]
When I run this version against the cnb fixture, I am getting a ruby error at the very end of the execution. 

```
$ bin/rundoc build --path=test/fixtures/cnb/ruby/rundoc.md

## Rundoc was successful, sanitizing output
## Writing RUNdoc output to /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240711-37789-qcibrm/README.md
## Running after build scripts
## Saving to /Users/stella.yang/Documents/rundoc/test/fixtures/cnb/ruby/project
## Cleaning working directory /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240711-37789-qcibrm
/Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `initialize': Permission denied @ rb_sysopen - /Users/stella.yang/Documents/rundoc/test/fixtures/cnb/ruby/project/./ruby-getting-started/.git/objects/pack/pack-f166dc4c4a9c5a7cf5d22d8c7887c625bc83560a.pack (Errno::EACCES)
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `open'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `block in copy_file'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2278:in `open'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2278:in `copy_file'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2243:in `copy'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1048:in `block in copy_entry'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2385:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1045:in `copy_entry'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:990:in `block in cp_r'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2469:in `block in fu_each_src_dest'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2483:in `fu_each_src_dest0'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2467:in `fu_each_src_dest'
	from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:989:in `cp_r'
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:168:in `copy_dir_contents'
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:158:in `on_success'
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:108:in `block in call'
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:96:in `chdir'
	from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:96:in `call'
	from bin/rundoc:72:in `build'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/command.rb:28:in `run'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor.rb:527:in `dispatch'
	from /Users/stella.yang/.gem/ruby/3.3.2/gems/thor-1.3.1/lib/thor/base.rb:584:in `start'
	from bin/rundoc:76:in `<main>'
```

File: README.md [hunk 1]
```diff
@@ -47,12 +47,12 @@ gem 'rundoc'
...
 
 ```sh
-$ bin/rundoc build --path <test/fixtures/rails_7/rundoc.md>
+$ rundoc build --path <test/fixtures/rails_7/rundoc.md>
 ```
 
 > Note: This command will create and manipulate directories in the working directory of your source markdown file. Best practice is to have your source markdown file in its own empty directory.
 
-This will generate a project folder with your project in it, and a markdown README.md with the parsed output of the markdown docs, and a copy of the source.
+This will generate a project folder with your project in it, and a markdown README.md with the parsed output of the markdown docs. See `rundoc --help` for more configuration options.
```
[the remark this answers is not in the document]
Comment (schneems) [reply README.md]: I prefer `--help`, it's more explicit and more likely to work with a greater range of commands I did add `rundoc help` though to print out help.

**Review by lilacstella — COMMENTED** [review]
```
$ bin/rundoc test/fixtures/cnb/ruby/rundoc.md --on-success-dir=yay --force
## Success! sanitizing output
## Writing RUNdoc output to /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240715-35761-q6esrf/README.md
## Running after build scripts 
## Saving to yay
## Cleaning working directory /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240715-35761-q6esrf
/Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2196:in `lstat': File name too long @ rb_file_s_lstat - /private/var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240715-35761-q6esrf/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/yay/ruby-getting-started/app/models/concerns/.keep (Errno::ENAMETOOLONG)
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2196:in `lstat'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2240:in `copy'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1048:in `block in copy_entry'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2385:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        ...
        ... a lot of times
        ...
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1045:in `copy_entry'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:990:in `block in cp_r'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2469:in `block in fu_each_src_dest'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2483:in `fu_each_src_dest0'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2467:in `fu_each_src_dest'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:989:in `cp_r'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:205:in `on_success'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:156:in `block in call'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:144:in `chdir'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:144:in `call'
        from bin/rundoc:19:in `<main>'
```

**Review by lilacstella — COMMENTED** [review]
```
$ bin/rundoc test/fixtures/cnb/ruby/rundoc.md 
## Success! sanitizing output
## Writing RUNdoc output to /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240715-42464-neu7b0/README.md
## Running after build scripts 
## Saving to /Users/stella.yang/Documents/rundoc/test/fixtures/cnb/ruby/project
## Cleaning working directory /var/folders/5n/dl9rnvh55t5bndgk2p_bxqx00000gn/T/d20240715-42464-neu7b0
/Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `initialize': Permission denied @ rb_sysopen - /Users/stella.yang/Documents/rundoc/test/fixtures/cnb/ruby/project/./ruby-getting-started/.git/objects/pack/pack-f166dc4c4a9c5a7cf5d22d8c7887c625bc83560a.pack (Errno::EACCES)
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `open'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2279:in `block in copy_file'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2278:in `open'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2278:in `copy_file'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2243:in `copy'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1048:in `block in copy_entry'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2385:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2388:in `block in wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `each'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2387:in `wrap_traverse'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:1045:in `copy_entry'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:990:in `block in cp_r'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2469:in `block in fu_each_src_dest'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2483:in `fu_each_src_dest0'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:2467:in `fu_each_src_dest'
        from /Users/stella.yang/.rubies/ruby-3.3.2/lib/ruby/3.3.0/fileutils.rb:989:in `cp_r'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:216:in `copy_dir_contents'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:206:in `on_success'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:156:in `block in call'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:144:in `chdir'
        from /Users/stella.yang/Documents/rundoc/lib/rundoc/cli.rb:144:in `call'
        from bin/rundoc:19:in `<main>'
```

**Comment by schneems** [comment]
Error was related to git and file permissions, this seems to fix it https://github.com/zombocom/rundoc/pull/58/commits/7babc8f7259b6f1b33adb402d43d40b82fd05128. Tests are passing again. Requesting re-review.

**Review by lilacstella — APPROVED** [review]

**Pull request merged**